### PR TITLE
Remove SCons

### PR DESCRIPTION
**Pull request by Holzhaus**

### CODE REVIEW

**Comment by Be-ing** [comment]
Thanks. We can't merge this until we get CMake working on macOS and switch the build servers to CMake.

File: build/debian/changelog
Comment (daschuer) [inline build/debian/changelog:1]: We need most of the diebisn files to create the deb packages.
I know that there is some automation with cmake and pebuilder, but some files are required anyway.

To keep almost empty changelog is stupid anyway. I think we need to find a way to make our main changelog file debian compliment.

**Review by daschuer — CHANGES_REQUESTED** [review]
I think most of the files in the debian folder need to remain.
We need to find out what is the standard way of creating a deb file making use of cmake and follow it.

File: build/debian/changelog
Comment (uklotzde) [reply build/debian/changelog:1]: Typical entry in the (RPMFusion) distribution changelog: "New upstream release". Otherwise only technical changes when bumping the .spec version.

This is a different kind of changelog.

Comment (uklotzde) [reply build/debian/changelog:1]: Not everyone uses a Debian system and the main changelog should stay agnostic.

Comment (Holzhaus) [reply build/debian/changelog:1]: You can already build a DEB package via `cpack -G DEB`.

**Comment by uklotzde** [comment]
I would like to enforce the removal of SCons:
- Keeping the SCons build consistent and working requires work for all new features.
- The SCons builds consume precious Travis computing time.
- If we don't do it there will be no pressure to complete the remaining tasks until 2.4 beta is imminent.

**Comment by Be-ing** [comment]
Unfortunately we still need someone to make CMake builds not crash on startup on macOS...

**Comment by uklotzde** [comment]
> Unfortunately we still need someone to make CMake builds not crash on startup on macOS...

No one seems to care as long as the SCons builds are available. That's my point.

**Comment by daschuer** [comment]
> No one seems to care as long as the SCons builds are available. That's my point.

But if there is simply no one we are running into dead end with no MacOs support. I think we should find a contributor fro MacOS CMake first.

**Comment by uklotzde** [comment]
Unfortunately finding volunteers for fixing macOS issues didn't work well in the past. I don't expect anyone will step in until we cut off this convenient workaround. In the worst case just a few weeks without nightly macOS builds.

**Comment by Be-ing** [comment]
Considering we have a usable 2.3 beta branch with macOS builds now, this may be a good time to do that.

**Comment by daschuer** [comment]
@crisclacerda has tried it unsuccessfully, he is now at Linux. But we will loose him entirely as master tester if we merge this without a working CMake.

**Comment by Be-ing** [comment]
I am afraid breaking master with macOS will have the opposite effect and discourage macOS contributors.

**Comment by ywwg** [comment]
Do we have an OS requirement for mac? I have an old (2009) Mac Pro that can run Mixxx, but I think the nvidia graphics card isn't supported in the newest OS

**Comment by Be-ing** [comment]
I think macOS 10.13 is our current minimum.

**Comment by ywwg** [comment]
it looks like that's the last OS that supports nvidia so I might be able to make it work.  I'll see if I can get it up and running

**Comment by Be-ing** [comment]
That would be great. Although I presume it will be painfully slow to keep retrying failing builds on a computer from 2009.

**Comment by ywwg** [comment]
it was a beefy machine in 2009 capable of realtime hi rez video editing, so it should still be able to chew through gcc easily

**Comment by jcelerier** [comment]
Hi, I'm willing to test this on macOS - can someone explain to me what exactly is the test procedure ? should I just get this PR, and check that it builds with cmake ?

**Comment by Be-ing** [comment]
@jcelerier thanks for your interest in helping. Please read the prior [Zulip discussion](https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/cmake.20on.20macOS/near/194884364) we had about the problem. There is a crash on startup in libsqlite3.dylib. I suspect this has to do with how CMake is linking Mixxx to SQLite and the custom collate function we use with SQLite.

**Comment by Be-ing** [comment]
@jcelerier I see you also have a lot of experience with QML and you've ported RtMidi to C++17. We are currently brainstorm a new JavaScript controller mapping system including how we could use QML for controller screens. If you're interested in helping with that, take a look at [this wiki page summarizing the current plans](https://github.com/mixxxdj/mixxx/wiki/New-controller-system) and the Zulip discussions linked from that page.

**Comment by Holzhaus** [comment]
@jcelerier You don't need to test this PR (it just deletes SCons), just try to compile the master branch using CMake on MacOS and check if it's usable. Apparently it crashes on MacOS, but I can't debug it without access to a Mac.

**Comment by Be-ing** [comment]
Now that CMake packaging and code signing is working on macOS, I think we have one last obstacle before we can get rid of scons. Currently the PPA is updated by a Jenkins job running `scons makeubuntu`. Can we replace that with a script on Travis or GitHub Actions?

**Comment by JoergAtGithub** [comment]
Please document the steps needed to build Mixxx under Windows, before you remove the documented way using scons: https://github.com/mixxxdj/mixxx/wiki/compiling-on-windows

**Comment by daschuer** [comment]
Since  #3411 Scons is broken.
Are the Mac issues solved?
Only the PPA upload missing?

**Comment by Be-ing** [comment]
Yes, macOS builds, packaging, code signing, and deployment are all working with CMake now.

I think it might be easier to start this over on the 2.3 branch than resolve conflicts.

**Comment by daschuer** [comment]
This cannot merged, because the debian folder needs to remain.
Please wait until this is cmake-ified.

How should the new directory layout look like?
cmake/debian
Originally I thought about rename the build folder to packaging or something. This can than hold also the RPM and Arch scripts.

**Comment by Holzhaus** [comment]
I'm rebasing on 2.3 and take care of the issues.

**Comment by Be-ing** [comment]
> Originally I thought about rename the build folder to packaging or something.

I like the idea to not put everything in the `cmake` folder. `packaging` would work well I think. We could move the build environment setup scripts from `tools` to `packaging`.

**Comment by daschuer** [comment]
> I'm rebasing on 2.3 and take care of the issues.

Please hold on until my PPA branch is ready. Otherwise this will conflict again.

**Comment by Holzhaus** [comment]
> > Originally I thought about rename the build folder to packaging or something.
> 
> I like the idea to not put everything in the `cmake` folder. `packaging` would work well I think. We could move the build environment setup scripts from `tools` to `packaging`.

Not sure about this. I think we already have enough top-level directories, and the setup scripts are not related to packaging. They are also needed for macos/windows users that just want to build Mixxx locally.

**Comment by Holzhaus** [comment]
@daschuer i moved the Debian directory to cmake/ instead of removing it. The merge conflicts should be neglible then, right?

**Comment by daschuer** [comment]
For my understanding, the cmake folder contains cmake helper files, that are as good as possible project independent. 
If we move the individual packaging files there as well, it can be supprising for contributors. 

Setup script at for my understanding a part of the package. IMHO In all cases, you have the binary, additional data, metadata and setup scripts that are more or less packaged to one achieve.

**Comment by Holzhaus** [comment]
I meant the buildenv setup scripts that not supposed to be packaged. They are in the `tools/` directory where they belong, because they are helper scripts for developers/testers to make building mixxx easier.

**Comment by Be-ing** [comment]
Finally, let's do this. @Holzhaus there are some conflicts now.

File: .github/workflows/build.yml
Comment (daschuer) [inline .github/workflows/build.yml:136]: This should be moved to packages or better moved to GitHub.
The cmake folder should IMHO be free from mixxx specific stuff.

File: .github/workflows/build.yml
Comment (daschuer) [inline .github/workflows/build.yml:185]: Do all jobs have a new working directory?
I know an approach where you use subdirectories for different builds.
I this relevant here.
I use for instance a build/Debug folder and a build/RelWithDebugInfo folder.

File: CMakeLists.txt
Comment (daschuer) [inline CMakeLists.txt:1105]: This should go to packaging/bundle or such.

**Comment by Holzhaus** [comment]
Did I mess up the merge? Something is wrong with the debian source packaging.

File: build/wix/Localization/README-Translations.md
Comment (daschuer) [inline build/wix/Localization/README-Translations.md:2]: I think we still need these files, right?

File: cmake/debian/mixxx-test.install
Comment (daschuer) [inline cmake/debian/mixxx-test.install:2]: The cmake debian folder can be removed

**Review by daschuer — CHANGES_REQUESTED** [review]
I have no idea why the source package export fails. The error message is not very helpful.

I suggest to start this over by rebasing.

File: .github/workflows/build.yml
Comment (Holzhaus) [reply .github/workflows/build.yml:185]: > Do all jobs have a new working directory?

On GitHub Actions, yes.

> I know an approach where you use subdirectories for different builds.
> I this relevant here.
> I use for instance a build/Debug folder and a build/RelWithDebugInfo folder

Every build starts in a clean container, so there is no reason to use different directories on GitHub Actions.

Comment (daschuer) [reply .github/workflows/build.yml:185]: Ok.

That's might be the issue for the packaging fail. Cpack alown fails.
You need to configure it with cmake first.

File: build/wix/Localization/README-Translations.md
Comment (Holzhaus) [reply build/wix/Localization/README-Translations.md:2]: No, localized WIX installers are a total hack and I wasted way too much time to make them work with CPack. If nobody volunteers to implement it, let's remove them. We could just pull them from transifex if somebody wants to add them, right?

File: .github/workflows/build.yml
Comment (Holzhaus) [reply .github/workflows/build.yml:185]: The reason for the packaging fail was that the PPA step still used the `cmake_build` directory instead of `build`, but since that step was new it didn't caused a conflict. I amended the merge commit.

File: build/wix/Localization/README-Translations.md
Comment (daschuer) [reply build/wix/Localization/README-Translations.md:2]: What is the current situation?
Scons = localized wix 
Cmake = english only?

**Comment by daschuer** [comment]
Can you rebase / squash commits to avoid double renamings In the commit history?

**Comment by Holzhaus** [comment]
Before merging, let's prepare a blog post, as this will probably break the workflow for a bunch of third party package maintainers that still use SCons (e. g. the `mixxx-git` package on the AUR). I'll write one this evening.

**Review by Be-ing — APPROVED** [review]

**Comment by Be-ing** [comment]
You can discuss moving to GitHub Actions from our old build servers in that post too.

**Comment by uklotzde** [comment]
> Before merging, let's prepare a blog post, as this will probably break the workflow for a bunch of third party package maintainers that still use SCons (e. g. the `mixxx-git` package on the AUR). I'll write one this evening.

Already happened: https://bugs.launchpad.net/mixxx/+bug/1907528

**Comment by Be-ing** [comment]
Good point. SCons is already broken. Merge now before waiting for the blog post?

**Comment by uklotzde** [comment]
> Good point. SCons is already broken. Merge now before waiting for the blog post?

No need to wait. It is already broken and main does not build with SCons.

**Comment by Be-ing** [comment]
We still need to update the Windows build documentation.

**Comment by Holzhaus** [comment]
https://github.com/mixxxdj/website/pull/171

**Pull request merged**
